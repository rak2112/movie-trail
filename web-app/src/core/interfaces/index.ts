type MovieType =
  | 'Action'
  | 'Adventure'
  | 'Animation'
  | 'Comedy'
  | 'Crime'
  | 'Documentry'
  | 'Drama'
  | 'Family'
  | 'Fantasy'
  | 'History'
  | 'Horror'
  | 'Music'
  | 'Mystery'
  | 'Romance'
  | 'Science Fiction'
  | 'TV Movie'
  | 'Thriller'
  | 'War'
  | 'Western';

export type UserMovieType= 'FAVORITE' | 'WATCHLIST';

export type UserView = 'Favorites' | 'Watchlist';

export interface Api {
  isFetching?: boolean;
  hasError?: boolean;
  errorMessage?: string;
}; 

export interface FormValues {
  name?: string;
  email: string;
  password: string;
  confirmPassword?: string;
  token?: string;
}

export interface Language {
  id: string;
  name: string;
}

export interface Movie {
  id: number;
  backdrop_path: string;
  genre_ids: number[];
  genres: Genre[];
  homepage: string;
  overview: string;
  poster_path?: string;
  runtime: string;
  release_date: string;
  spoken_languages: Language[];
  status: string;
  title: string;
  type?: string;
  vote_average: string;
};

export interface Movies {
  pageNo: number;
  totalPages: number;
  totalResults: number;
  results: Movie[];
  redirection? : boolean;
};

export interface MoviesData {
  pageNo: number;
  total_results: number;
  total_pages: number;
  results?: Movie[];
};

export interface MoviesRes {
  movies: Movie[]
};

export interface Genre {
  id: number;
  name: MovieType;
};

export interface GenresMap {
  [id: number]: Genre;
};

export interface UserMovie extends Movie {
  _id?: string;
  movieType?: UserMovieType;
}

export interface User {
  email?: string;
  displayName?: string;
  loggedIn?: boolean;
  movies?: UserMovie[];
  resetMessage?: string;
  token?: string;
}

export interface UserMovies {
  [id: string]: UserMovie;
}
export interface UserMovieMap {
  favorites: UserMovies;
  watchlist: UserMovies;
}

export interface Cast {
  character: string;
  id: string;
  name: string;
  profile_path?: string;
}

export interface Crew extends Cast {
  department: string;
  job: string;
}

export interface Poster {
  file_path: string;
}

export interface Persons {
  id?: string;
  cast: Cast[];
  crew: Crew[];
}

export interface Posters {
  id?: string;
  backdrops: Poster[];
  posters: Poster[];
}

export interface Video {
  id: string;
  key: string;
  name: string;
}

export interface Videos {
  id: string;
  results: Video[]
}

export interface MovieDetail {
  isFetching?: boolean;
  hasError?: boolean;
  details: Movie;
  images: Posters;
  persons: Persons;
  videos: Videos
}