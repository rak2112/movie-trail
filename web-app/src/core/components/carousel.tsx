import React, { PureComponent } from 'react';
import styled from 'react-emotion';

import { ImageLoader } from '../components';
import { MovieDetail, Poster, Posters } from '../interfaces';

import { paths } from '../utils';

export const Container = styled('div')`
  img.poster {
    width: 550px;
    height: 400px;
    border: 4px solid #222;
    margin: 0 30px;
  }
  .thumbnails {
    padding-top: 20px;
    text-align: center;
    .movie {
      display: inline-block;
      margin: 0 5px;
      cursor: pointer;
      width: 55px;
      height: 78px;
    }
    .active {
      border: 1px solid #ff5c00;
    }
  }
`;

interface State {
  active: number;
  images: Poster[];
}

export interface Props {
  images: Posters;
};

export class Carousel extends PureComponent <Props, State>{

  static displayName = 'CarouselComponent';
  static getDerivedStateFromProps({ images: { posters } }: MovieDetail) {
    const images = posters.slice(0, 5);
    return { images };
  }

  state = {
    images: [],
    active: 1
  };

  handleClick = (event: any) => {
    this.setState({active: +event.target.dataset.index});
  }

  render() {
    const { active, images }: State = this.state;
    return (
      <Container>
        <img className="poster" src={`${paths.imgPath500}${images[active].file_path}`} alt=""/>
        <div className="thumbnails">
          {
            images.map((img: Poster, index: number) =>
              <ImageLoader 
                className={ index === active ? 'active thumbnail movie' : 'thumbnail movie'}
                alt={`movie thumbnail ${index}`}
                clickHandler={this.handleClick}
                key={img.file_path}
                index={index}
                src={`${paths.imgPath45}${img.file_path}`}
              />
            )
          }
        </div>
      </Container>
    )
  }
}