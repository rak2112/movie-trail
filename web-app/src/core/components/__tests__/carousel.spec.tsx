import { shallow } from 'enzyme';
import React, { Component } from "react";
import { create } from 'react-test-renderer';
import { Poster } from '../../interfaces';
import { Carousel, Props } from '../carousel';

describe(`Carousel Component`, () => {

  const props = {
    handleClick: ()=> {return},
    images: {
      backdrops: [
        {
          file_path: '/backdrop/1'
        } as Poster
      ],
      posters: [{
        file_path: '/path/1'
      } as Poster,
      {
        file_path: '/path/2'
      } as Poster,
      {
        file_path: '/path/3'
      } as Poster]
    }
  };

  it(`should match snapshot`, () => {
    const wrapper = create(
      <Carousel {...props}/>
    ).toJSON();
    
    expect(wrapper).toMatchSnapshot();
  });
  
  it(`should render with desired props`, () => {
    const wrapper = shallow(
      <Carousel {...props}/>
    );
    const instance = wrapper.instance() as Carousel;
    
    expect(instance.state.active).toBe(1);
    expect(instance.state.images.length).toBe(3);
    expect(wrapper.find('.poster').prop('src')).toContain('/path/2');
    expect(wrapper.find('.thumbnails').children().length).toEqual(3);

    instance.handleClick({ target: { dataset: { index: 2 }}});
    expect(instance.state.active).toBe(2);
    expect(wrapper.find('.poster').prop('src')).toContain('/path/3');
  });

});