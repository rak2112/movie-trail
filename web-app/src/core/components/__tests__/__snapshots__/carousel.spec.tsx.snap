// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Component should match snapshot 1`] = `
<div
  className="css-45larf"
>
  <img
    alt=""
    className="poster"
    src="https://image.tmdb.org/t/p/w500/path/2"
  />
  <div
    className="thumbnails"
  >
    <img
      alt="movie thumbnail 0"
      className="thumbnail movie"
      src="loader.gif"
      style={undefined}
    />
    <img
      alt="movie thumbnail 1"
      className="active thumbnail movie"
      src="loader.gif"
      style={undefined}
    />
    <img
      alt="movie thumbnail 2"
      className="thumbnail movie"
      src="loader.gif"
      style={undefined}
    />
  </div>
</div>
`;
